# The Battle of Neighborhoods
## Coursera - Applied Data Science Capstone Project

### Project Description
I love pizza!  I like deep dish, New York style, thin crust and even Pizza rolls.  COVID-19 has changed by work life and I can now move to any city I want as long as I have a good internet connection.  I want to find a new medium sized city with the most Pizza places per capita.  I will then try to find a spot in that city to look for my new home.

### Description of the Data
#### Where is the Data Located?
The Data for my project is located in two sources.  The first is Wikipedia for the citys in the US. It contains a list of the largest 300 cities in the use.  I will use FourSquare to find the number and location of the pizza restaurants around each city.

#### How will the data be used to solve the problem?
The data will be used to first find a list of medium cities.  I will filter that long list of citys to a subset of medium sized cities (250K < Pop <1M).  After I have a shorter list of cities, I will then look for the number of pizza places within 5 km of city center.  I will then find the per Capita number of Pizza locations per city.  I will then select the city with the most Pizza restaraunts per capita and plot the locations.  After I have a general idea where they are located I will cluster the list of Pizza shops into groups.  After I have clustered the locations, I will find the group with the highest number of locations in the cluster and start looking for a house around that centroid.

## Step 1. Scrape Wikipedia for the list of Neighborhoods

In [2]:
# Import Key Libraries into the Notebook
import pandas as pd
import numpy as np

In [3]:
import requests # library to handle requests
import random # library for random number generation


In [4]:
# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize


In [5]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    ------------------------------------------------------------
                       

In [6]:
# module to convert an address into latitude and longitude values
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim 
print('GeoCoders installed')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.50-py_0          conda-forge
    geopy:         2.0.0-pyh9f0ad1d_0 conda-forge


geopy-2.0.0          | 63 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
GeoCoders installed


In [7]:
# import the library we use to open URLs
import urllib.request as URLRequest

### Now that all the libraries are Imported, We can move on to working with data

### Read in the web address of US Cities by Population

In [8]:
# specify which URL/web page we are going to be scraping
url = "https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population"

In [9]:
# open the url using urllib.request and put the HTML into the page variable
page = URLRequest.urlopen(url)

In [10]:
# import the BeautifulSoup library so we can parse HTML and XML documents
from bs4 import BeautifulSoup

In [11]:
# parse the HTML from our URL into the BeautifulSoup parse tree format
soup = BeautifulSoup(page, "lxml")

In [12]:
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of United States cities by population - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"04059fbf-36f9-4119-8245-e104660d26c5","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_United_States_cities_by_population","wgTitle":"List of United States cities by population","wgCurRevisionId":961764896,"wgRevisionId":961764896,"wgArticleId":1649321,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Wikipedia indefinitely semi-pr

In [13]:
# use the 'find_all' function to bring back all instances of the 'table' tag in the HTML and store in 'all_tables' variable
all_tables=soup.find_all("table")
all_tables

[<table class="noprint infobox" id="GeoGroup" style="width: 23em; font-size: 88%; line-height: 1.5em">
 <tbody><tr>
 <td><b>Map all coordinates using:</b> <a class="external text" href="//tools.wmflabs.org/osm4wiki/cgi-bin/wiki/wiki-osm.pl?project=en&amp;article=List_of_United_States_cities_by_population">OpenStreetMap</a> 
 </td></tr>
 <tr>
 <td><b>Download coordinates as:</b> <a class="external text" href="//tools.wmflabs.org/kmlexport?article=List_of_United_States_cities_by_population">KML</a> <b>·</b> <a class="external text" href="http://tripgang.com/kml2gpx/http%3A%2F%2Ftools.wmflabs.org%2Fkmlexport%3Farticle%3DList_of_United_States_cities_by_population?gpx=1" rel="nofollow">GPX</a>
 </td></tr></tbody></table>,
 <table class="vertical-navbox nowraplinks navbox" style="float:right;clear:right;width:22.0em;margin:0 0 1.0em 1.0em;background:#f8f9fa;border:1px solid #aaa;padding:0.2em;border-spacing:0.4em 0;text-align:center;line-height:1.4em;font-size:88%"><tbody><tr><th style="padd

In [14]:
right_table=soup.find('table', class_='wikitable sortable')
right_table

<table class="wikitable sortable" style="text-align:center">
<tbody><tr>
<th>2019<br/>rank
</th>
<th>City
</th>
<th>State<sup class="reference" id="cite_ref-5"><a href="#cite_note-5">[c]</a></sup>
</th>
<th>2019<br/>estimate
</th>
<th>2010<br/>Census
</th>
<th>Change
</th>
<th colspan="2">2016 land area
</th>
<th colspan="2">2016 population density
</th>
<th>Location
</th></tr>
<tr>
<td>1
</td>
<td style="text-align:left;background-color:#cfecec"><i><a href="/wiki/New_York_City" title="New York City">New York</a></i><sup class="reference" id="cite_ref-6"><a href="#cite_note-6">[d]</a></sup>
</td>
<td align="left"><span class="flagicon"><img alt="" class="thumbborder" data-file-height="1137" data-file-width="2276" decoding="async" height="11" src="//upload.wikimedia.org/wikipedia/commons/thumb/1/1a/Flag_of_New_York.svg/23px-Flag_of_New_York.svg.png" srcset="//upload.wikimedia.org/wikipedia/commons/thumb/1/1a/Flag_of_New_York.svg/35px-Flag_of_New_York.svg.png 1.5x, //upload.wikimedia.org

### Read the HTML into a dataset and add to the Pandas Datafrome

In [15]:
# The data is spread across 11 columns so grab all data and keep what is important
A=[]
B=[]
C=[]
D=[]
E=[]
F=[]
G=[]
H=[]
I=[]
J=[]
K=[]
L=[]
for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==11:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))
        D.append(cells[3].find(text=True))
        E.append(cells[4].find(text=True))
        F.append(cells[5].find(text=True))
        H.append(cells[6].find(text=True))
        I.append(cells[7].find(text=True))
        J.append(cells[8].find(text=True))
        K.append(cells[9].find(text=True))
        L.append(cells[10].find(text=True))
        
        
                

### This is a full list of potential cities

In [16]:

# Create the Dataframe and assign the array for each field to a column
df=pd.DataFrame(A,columns=['Number'])
df['CityName']=B
df['Population']=D
df

Number        CityName  Population
0       1
        New York  8,336,817

1       2
     Los Angeles  3,979,576

2       3
         Chicago  2,693,976

3       4
         Houston  2,320,268

4       5
         Phoenix  1,680,992

5       6
    Philadelphia  1,584,064

6       7
     San Antonio  1,547,253

7       8
       San Diego  1,423,851

8       9
          Dallas  1,343,573

9      10
        San Jose  1,021,795

10     11
          Austin    978,908

11     12
    Jacksonville    911,507

12     13
      Fort Worth    909,585

13     14
        Columbus    898,553

14     15
       Charlotte    885,708

15     16
   San Francisco    881,549

16     17
    Indianapolis    876,384

17     18
         Seattle    753,675

18     19
          Denver    727,211

19     20
      Washington    705,749

20     21
          Boston    692,600

21     22
         El Paso    681,728

22     23
       Nashville    670,820

23     24
         Detroit    670,031

24     25
   Oklahoma City    655,057

25     26
        Portland    654,741

26     27
       Las Vegas    651,319

27     28
         Memphis    651,073

28     29
      Louisville    617,638

29     30
       Baltimore    593,490

..     ...             ...         ...
286   288
           Tyler    106,985

287   289
           Davie    106,306

288   290
        Lakewood    106,300

289   291
       Daly City    106,280

290   292
         Boulder    105,673

291   293
           Allen    105,623

292   294
     West Covina    105,101

293   295
          Sparks    105,006

294   296
   Wichita Falls    104,683

295   297
       Green Bay    104,578

296   298
       San Mateo    104,430

297   299
         Norwalk    103,949

298   300
          Rialto    103,526

299   301
      Las Cruces    103,432

300   302
           Chico    103,301

301   303
        El Cajon    102,708

302   304
         Burbank    102,511

303   305
      South Bend    102,026

304   306
          Renton    101,751

305   307
           Vista    101,638

306   308
       Davenport    101,590

307   309
        Edinburg    101,170

308   310
      Tuscaloosa    101,129

309   311
          Carmel    101,068

310   312
  Spokane Valley    101,060

311   313
      San Angelo    101,004

312   314
       Vacaville    100,670

313   315
         Clinton    100,471

314   316
            Bend    100,421

315   317
      Woodbridge    100,145


[316 rows x 3 columns]

In [17]:
# Check the data types before I filter the cities for size
df.dtypes


Number        object
CityName      object
Population    object
dtype: object

## Step 2. Filter the list down to medium sized cities

### Now that I have a list of US cities, I need to convert from object to a number 

In [18]:
# Remove the commas from the numbers, and remove the '\n' from the end of every number
df['Population']=df['Population'].str.replace(',','')
df['Population']=df['Population'].str.replace('\n','')

In [19]:
# Convert the data type from Object to Float
df["Population"] = df['Population'].astype(float)

### Filter out Cities that are too small or too large

In [20]:
# Remove all cities with a population less than 250,000
df1 = df.loc[df['Population'] > 250000]
df1

Number         CityName  Population
0      1
         New York   8336817.0
1      2
      Los Angeles   3979576.0
2      3
          Chicago   2693976.0
3      4
          Houston   2320268.0
4      5
          Phoenix   1680992.0
5      6
     Philadelphia   1584064.0
6      7
      San Antonio   1547253.0
7      8
        San Diego   1423851.0
8      9
           Dallas   1343573.0
9     10
         San Jose   1021795.0
10    11
           Austin    978908.0
11    12
     Jacksonville    911507.0
12    13
       Fort Worth    909585.0
13    14
         Columbus    898553.0
14    15
        Charlotte    885708.0
15    16
    San Francisco    881549.0
16    17
     Indianapolis    876384.0
17    18
          Seattle    753675.0
18    19
           Denver    727211.0
19    20
       Washington    705749.0
20    21
           Boston    692600.0
21    22
          El Paso    681728.0
22    23
        Nashville    670820.0
23    24
          Detroit    670031.0
24    25
    Oklahoma City    655057.0
25    26
         Portland    654741.0
26    27
        Las Vegas    651319.0
27    28
          Memphis    651073.0
28    29
       Louisville    617638.0
29    30
        Baltimore    593490.0
..    ...              ...         ...
59    60
        Lexington    323152.0
60    61
        Henderson    320189.0
61    62
         Stockton    312697.0
62    63
       Saint Paul    308096.0
63    64
       Cincinnati    303940.0
64    65
        St. Louis    300576.0
65    66
       Pittsburgh    300286.0
66    67
       Greensboro    296710.0
67    68
          Lincoln    289102.0
68    69
        Anchorage    288000.0
69    70
            Plano    287677.0
70    71
          Orlando    287442.0
71    72
           Irvine    287401.0
72    73
           Newark    282011.0
73    74
           Durham    278993.0
74    75
      Chula Vista    274492.0
75    76
           Toledo    272779.0
76    77
       Fort Wayne    270402.0
77    78
   St. Petersburg    265351.0
78    79
           Laredo    262491.0
79    80
      Jersey City    262075.0
80    81
         Chandler    261165.0
81    82
          Madison    259680.0
82    83
          Lubbock    258862.0
83    84
       Scottsdale    258069.0
84    85
             Reno    255601.0
85    86
          Buffalo    255284.0
86    87
          Gilbert    254114.0
87    88
         Glendale    252381.0
88    89
  North Las Vegas    251974.0

[89 rows x 3 columns]

In [21]:
# Remove the cities with a population of greater than 1 million from the list
dfMediumCity = df1.loc[df['Population'] < 1000000]
dfMediumCity

Number          CityName  Population
10    11
            Austin    978908.0
11    12
      Jacksonville    911507.0
12    13
        Fort Worth    909585.0
13    14
          Columbus    898553.0
14    15
         Charlotte    885708.0
15    16
     San Francisco    881549.0
16    17
      Indianapolis    876384.0
17    18
           Seattle    753675.0
18    19
            Denver    727211.0
19    20
        Washington    705749.0
20    21
            Boston    692600.0
21    22
           El Paso    681728.0
22    23
         Nashville    670820.0
23    24
           Detroit    670031.0
24    25
     Oklahoma City    655057.0
25    26
          Portland    654741.0
26    27
         Las Vegas    651319.0
27    28
           Memphis    651073.0
28    29
        Louisville    617638.0
29    30
         Baltimore    593490.0
30    31
         Milwaukee    590157.0
31    32
       Albuquerque    560513.0
32    33
            Tucson    548073.0
33    34
            Fresno    531576.0
34    35
              Mesa    518012.0
35    36
        Sacramento    513624.0
36    37
           Atlanta    506811.0
37    38
       Kansas City    495327.0
38    39
  Colorado Springs    478221.0
39    40
             Omaha    478192.0
..    ...               ...         ...
59    60
         Lexington    323152.0
60    61
         Henderson    320189.0
61    62
          Stockton    312697.0
62    63
        Saint Paul    308096.0
63    64
        Cincinnati    303940.0
64    65
         St. Louis    300576.0
65    66
        Pittsburgh    300286.0
66    67
        Greensboro    296710.0
67    68
           Lincoln    289102.0
68    69
         Anchorage    288000.0
69    70
             Plano    287677.0
70    71
           Orlando    287442.0
71    72
            Irvine    287401.0
72    73
            Newark    282011.0
73    74
            Durham    278993.0
74    75
       Chula Vista    274492.0
75    76
            Toledo    272779.0
76    77
        Fort Wayne    270402.0
77    78
    St. Petersburg    265351.0
78    79
            Laredo    262491.0
79    80
       Jersey City    262075.0
80    81
          Chandler    261165.0
81    82
           Madison    259680.0
82    83
           Lubbock    258862.0
83    84
        Scottsdale    258069.0
84    85
              Reno    255601.0
85    86
           Buffalo    255284.0
86    87
           Gilbert    254114.0
87    88
          Glendale    252381.0
88    89
   North Las Vegas    251974.0

[79 rows x 3 columns]

### Now that I have a list of potential medium sized cities, I will look up their Coordinates for each city

In [22]:
# Loop through the rows in the dataframe and request the location.
Lat=[]
Long=[]

for index, row in dfMediumCity.iterrows():
    address= row['CityName']
    geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(address)
    # Break the Location data into Latitude and Longitude
    latitude = location.latitude
    longitude = location.longitude
    Lat.append(latitude)
    Long.append(longitude)

In [23]:
# Update the Dataframe with the Latitude and Longitude of every city
dfMediumCity['Latitude']=Lat
dfMediumCity['Longitude']=Long
dfMediumCity

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.launch_new_instance()


Number          CityName  Population   Latitude   Longitude
10    11
            Austin    978908.0  30.271129  -97.743700
11    12
      Jacksonville    911507.0  30.332184  -81.655651
12    13
        Fort Worth    909585.0  32.753177  -97.332746
13    14
          Columbus    898553.0  39.962260  -83.000707
14    15
         Charlotte    885708.0  35.227200  -80.843083
15    16
     San Francisco    881549.0  37.779026 -122.419906
16    17
      Indianapolis    876384.0  39.768333  -86.158350
17    18
           Seattle    753675.0  47.603832 -122.330062
18    19
            Denver    727211.0  39.739236 -104.984862
19    20
        Washington    705749.0  38.894985  -77.036571
20    21
            Boston    692600.0  42.360253  -71.058291
21    22
           El Paso    681728.0  31.775415 -106.464635
22    23
         Nashville    670820.0  36.162230  -86.774353
23    24
           Detroit    670031.0  42.331551  -83.046640
24    25
     Oklahoma City    655057.0  35.472989  -97.517054
25    26
          Portland    654741.0  45.520247 -122.674195
26    27
         Las Vegas    651319.0  36.167256 -115.148516
27    28
           Memphis    651073.0  35.149022  -90.051629
28    29
        Louisville    617638.0  38.254238  -85.759407
29    30
         Baltimore    593490.0  39.290882  -76.610759
30    31
         Milwaukee    590157.0  43.034993  -87.922497
31    32
       Albuquerque    560513.0  35.084103 -106.650985
32    33
            Tucson    548073.0  32.222876 -110.974848
33    34
            Fresno    531576.0  36.729529 -119.708861
34    35
              Mesa    518012.0  33.415112 -111.831477
35    36
        Sacramento    513624.0  38.581572 -121.494400
36    37
           Atlanta    506811.0  33.749099  -84.390185
37    38
       Kansas City    495327.0  39.100105  -94.578142
38    39
  Colorado Springs    478221.0  38.833958 -104.825349
39    40
             Omaha    478192.0  41.258746  -95.938376
..    ...               ...         ...        ...         ...
59    60
         Lexington    323152.0  38.046407  -84.497039
60    61
         Henderson    320189.0  32.182598  -95.789318
61    62
          Stockton    312697.0  37.957702 -121.290780
62    63
        Saint Paul    308096.0  49.431805    2.010002
63    64
        Cincinnati    303940.0  39.101454  -84.512460
64    65
         St. Louis    300576.0  38.626804  -90.199410
65    66
        Pittsburgh    300286.0  40.441694  -79.990086
66    67
        Greensboro    296710.0  36.072635  -79.791975
67    68
           Lincoln    289102.0  47.829907 -118.414727
68    69
         Anchorage    288000.0  61.216313 -149.894852
69    70
             Plano    287677.0  33.013676  -96.692510
70    71
           Orlando    287442.0  28.542111  -81.379030
71    72
            Irvine    287401.0  33.685697 -117.825982
72    73
            Newark    282011.0  40.735657  -74.172367
73    74
            Durham    278993.0  35.996653  -78.901805
74    75
       Chula Vista    274492.0  32.640054 -117.084196
75    76
            Toledo    272779.0  39.856068   -4.023957
76    77
        Fort Wayne    270402.0  41.079990  -85.138601
77    78
    St. Petersburg    265351.0  59.960674   30.158655
78    79
            Laredo    262491.0  27.519984  -99.495376
79    80
       Jersey City    262075.0  40.728158  -74.077642
80    81
          Chandler    261165.0  33.306222 -111.841332
81    82
           Madison    259680.0  43.074761  -89.383761
82    83
           Lubbock    258862.0  33.563521 -101.879336
83    84
        Scottsdale    258069.0  33.494219 -111.926018
84    85
              Reno    255601.0  39.526121 -119.812658
85    86
           Buffalo    255284.0  42.886717  -78.878392
86    87
           Gilbert    254114.0  33.352763 -111.789037
87    88
          Glendale    252381.0  34.146942 -118.247847
88    89
   North Las Vegas    251974.0  36.200584 -115.121584

[79 rows x 5 columns]

## Step 3. Find the number of Pizza Locations in a 5km circle around the city

### Now that I have a great list of potential cities, I need to find out which ones have Pizza

In [24]:
# The code was removed by Watson Studio for sharing.

In [25]:
latitude = '38.254238'
longitude = '-85.759407'

In [26]:
search_query = 'Pizza'
# Radius of 5 km around city center
radius = 50000

In [27]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=VCDOEWVG3P3ALVKGAOQIC5MPMCQ01FT5HY35MK1B01SW4TCC&client_secret=IJSCBFL1KMKV3PWK5W4O0JJ0OKNFGRL0HSA22T5LW0UVFVJ4&ll=38.254238,-85.759407&v=20180604&query=Pizza&radius=50000&limit=30'

In [28]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5efb0803564b600a12d4578c'},
 'response': {'venues': [{'id': '4b5b7066f964a52038fe28e3',
    'name': 'Sicilian Pizza & Pasta',
    'location': {'address': '631 S 4th St',
     'lat': 38.24796985389662,
     'lng': -85.75802815295759,
     'labeledLatLngs': [{'label': 'display',
       'lat': 38.24796985389662,
       'lng': -85.75802815295759},
      {'label': 'entrance', 'lat': 38.247937, 'lng': -85.75796}],
     'distance': 708,
     'postalCode': '40202',
     'cc': 'US',
     'city': 'Louisville',
     'state': 'KY',
     'country': 'United States',
     'formattedAddress': ['631 S 4th St',
      'Louisville, KY 40202',
      'United States']},
    'categories': [{'id': '4bf58dd8d48988d1ca941735',
      'name': 'Pizza Place',
      'pluralName': 'Pizza Places',
      'shortName': 'Pizza',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
       'suffix': '.png'},
      'primary': True}],
    'delivery': {'id': '178816

In [29]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories delivery.id  \
0  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...     1788162   
1  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...         NaN   
2  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...         NaN   
3  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...         NaN   
4  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...         NaN   

               delivery.provider.icon.name  \
0  /delivery_provider_grubhub_20180129.png   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   

              delivery.provider.icon.prefix delivery.provider.icon.sizes  \
0  https://fastly.4sqi.net/img/general/cap/                     [40, 50]   
1                                       NaN                          NaN   
2                                       NaN                          NaN   
3                                       NaN                          NaN   
4                                       NaN                          NaN   

  delivery.provider.name                                       delivery.url  \
0                grubhub  https://www.grubhub.com/restaurant/sicilian-pi...   
1                    NaN                                                NaN   
2                    NaN                                                NaN   
3                    NaN                                                NaN   
4                    NaN                                                NaN   

   hasPerk                        id location.address  ...  \
0    False  4b5b7066f964a52038fe28e3     631 S 4th St  ...   
1    False  4c8edaebd68c6dcbb4eeffa1    110 W Main St  ...   
2    False  5b9aa75673fe25002c39fafc     445 S 4th St  ...   
3    False  4f3272b219836c91c7d8cac6   801 W Broadway  ...   
4    False  56cf4080498e5d8a7abb95d1              NaN  ...   

  location.crossStreet location.distance  \
0                  NaN               708   
1               2nd St               646   
2                  NaN               310   
3                  NaN               951   
4                  NaN               334   

                           location.formattedAddress  \
0  [631 S 4th St, Louisville, KY 40202, United St...   
1  [110 W Main St (2nd St), Louisville, KY 40202,...   
2  [445 S 4th St, Louisville, KY 40202, United St...   
3  [801 W Broadway, Louisville, KY 40202, United ...   
4                    [Louisville, KY, United States]   

                             location.labeledLatLngs  location.lat  \
0  [{'label': 'display', 'lat': 38.24796985389662...     38.247970   
1  [{'label': 'display', 'lat': 38.25605521414837...     38.256055   
2  [{'label': 'display', 'lat': 38.252133, 'lng':...     38.252133   
3  [{'label': 'display', 'lat': 38.2471809387207,...     38.247181   
4  [{'label': 'display', 'lat': 38.25169461774938...     38.251695   

  location.lng location.postalCode  location.state                     name  \
0   -85.758028               40202              KY   Sicilian Pizza & Pasta   
1   -85.752387               40202              KY     Impellizzeri's Pizza   
2   -85.757084               40202              KY                Pizza Bar   
3   -85.765556               40202              KY                Pizza Hut   
4   -85.757372                 NaN              KY  Birracibo Artisan Pizza   

     referralId  
0  v-1593509670  
1  v-1593509670  
2  v-1593509670  
3  v-1593509670  
4  v-1593509670  

[5 rows x 23 columns]

In [30]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name          categories  \
0                Sicilian Pizza & Pasta         Pizza Place   
1                  Impellizzeri's Pizza         Pizza Place   
2                             Pizza Bar         Pizza Place   
3                             Pizza Hut         Pizza Place   
4               Birracibo Artisan Pizza  Italian Restaurant   
5             Wick's Pizza Parlor & Pub         Pizza Place   
6                     Papa John's Pizza         Pizza Place   
7                             Pizza Hut         Pizza Place   
8                  Pizza Today Magazine     Coworking Space   
9                     Papa John's Pizza         Pizza Place   
10               DiOrio's Pizza and Pub         Pizza Place   
11                    Mr. Gatti's Pizza         Pizza Place   
12  Papa Murphy's | Take 'N' Bake Pizza         Pizza Place   
13                            Pizza Hut         Pizza Place   
14                 Impellizzeri's Pizza         Pizza Place   
15                    Papa John's Pizza         Pizza Place   
16                       Bearno's Pizza         Pizza Place   
17                       Domino's Pizza         Pizza Place   
18                    Mr. Gatti's Pizza         Pizza Place   
19                       Hometown Pizza         Pizza Place   
20                            Pizza Hut         Pizza Place   
21                            Pizza Hut         Pizza Place   
22                            Pizza Hut         Pizza Place   
23            BoomBozz Pizza & Taphouse         Pizza Place   
24                     Oak Street Pizza         Pizza Place   
25             Old Louisville Pizza Co.       Deli / Bodega   
26                           Debs Pizza         Pizza Place   
27                 Little Caesars Pizza         Pizza Place   
28                    Papa John's Pizza         Pizza Place   
29                       Domino's Pizza         Pizza Place   

                        address  cc            city        country  \
0                  631 S 4th St  US      Louisville  United States   
1                 110 W Main St  US      Louisville  United States   
2                  445 S 4th St  US      Louisville  United States   
3                801 W Broadway  US      Louisville  United States   
4                           NaN  US      Louisville  United States   
5                975 Baxter Ave  US      Louisville  United States   
6              117 W Liberty St  US      Louisville  United States   
7              2508 Preston Hwy  US      Louisville  United States   
8                           NaN  US      Louisville  United States   
9                 401 E Main St  US      Louisville  United States   
10               919 Baxter Ave  US      Louisville  United States   
11            2247 S Preston St  US      Louisville  United States   
12                3024 S 3rd St  US      Louisville  United States   
13   1020 Veterans Pkwy Ste 500  US     Clarksville  United States   
14            1381 Bardstown Rd  US      Louisville  United States   
15              2336 S Floyd St  US      Louisville  United States   
16        3002 Charlestown Xing  US      New Albany  United States   
17        2784 Meijer Dr Ste 1A  US  Jeffersonville  United States   
18  703 E Lewis and Clark Pkwy.  US     Clarksville  United States   
19             4041 Preston Hwy  US      Louisville  United States   
20              2733 W Broadway  US      Louisville  United States   
21              1817 S Brook St  US      Louisville  United States   
22          4211 Shelbyville Rd  US      Louisville  United States   
23            1448 Bardstown Rd  US      Louisville  United States   
24                 125 E Oak St  US      Louisville  United States   
25                          NaN  US      Louisville  United States   
26              2101 W Broadway  US      Louisville  United States   
27             5622 Preston Hwy  US      Louisville  United States   
28              2111 W Broadway  US      Louisville

### Now that I know that I can pull the Data for one city, I need to pull a full list for every city

In [31]:
dfPizza=pd.DataFrame(columns=list(dataframe_filtered.columns))
dfPizza

Empty DataFrame
Columns: [name, categories, address, cc, city, country, crossStreet, distance, formattedAddress, labeledLatLngs, lat, lng, postalCode, state, id]
Index: []

In [49]:
LIMIT = 500
fullresults=''

for index, row in dfMediumCity.iterrows():
    latitude= row['Latitude']
    longitude= row['Longitude']
    url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
    results = requests.get(url).json()
    # assign relevant part of JSON to venues
    venues = results['response']['venues']
# tranform venues into a dataframe
    dataframe = json_normalize(venues)
    dfPizza=pd.concat([dfPizza, dataframe], ignore_index=True)

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:13: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.



In [50]:
dfPizza.shape

(6749, 37)

In [94]:
dfPizza.head()

address                                         categories   cc city  \
0     NaN  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  NaN  NaN   
1     NaN  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  NaN  NaN   
2     NaN  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  NaN  NaN   
3     NaN  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  NaN  NaN   
4     NaN  [{'id': '4bf58dd8d48988d1ca941735', 'name': 'P...  NaN  NaN   

  country crossStreet delivery.id              delivery.provider.icon.name  \
0     NaN         NaN         NaN                                      NaN   
1     NaN         NaN         NaN                                      NaN   
2     NaN         NaN     1577785  /delivery_provider_grubhub_20180129.png   
3     NaN         NaN         NaN                                      NaN   
4     NaN         NaN         NaN                                      NaN   

              delivery.provider.icon.prefix delivery.provider.icon.sizes  ...  \
0                                       NaN                          NaN  ...   
1                                       NaN                          NaN  ...   
2  https://fastly.4sqi.net/img/general/cap/                     [40, 50]  ...   
3                                       NaN                          NaN  ...   
4                                       NaN                          NaN  ...   

  location.lat location.lng location.neighborhood location.postalCode  \
0    30.267108   -97.739088                   NaN               78701   
1    30.269797   -97.749849                   NaN               78701   
2    30.277757   -97.744391                   NaN               78701   
3    30.256824   -97.739480                   NaN               78701   
4    30.267732   -97.740302                   NaN               78701   

  location.state                 name postalCode    referralId state  \
0             TX   Jack Pizza Company        NaN  v-1593509670   NaN   
1             TX        Gatti's Pizza        NaN  v-1593509670   NaN   
2             TX      iFratelli Pizza        NaN  v-1593509670   NaN   
3             TX      Salvation Pizza        NaN  v-1593509670   NaN   
4             TX  Pizzaria Papparazzi        NaN  v-1593509670   NaN   

  venuePage.id  
0          NaN  
1     66438817  
2          NaN  
3    148383145  
4          NaN  

[5 rows x 37 columns]

In [95]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dfPizza_filtered = dfPizza.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dfPizza_filtered['categories'] = dfPizza_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dfPizza_filtered.columns = [column.split('.')[-1] for column in dfPizza_filtered.columns]

dfPizza_filtered

name            categories  \
0                    Jack Pizza Company           Pizza Place   
1                         Gatti's Pizza           Pizza Place   
2                       iFratelli Pizza           Pizza Place   
3                       Salvation Pizza           Pizza Place   
4                   Pizzaria Papparazzi           Pizza Place   
5                           Pie95 Pizza           Pizza Place   
6       Chicago Pizza and Sports Grille           Pizza Place   
7                          Paddys Pizza           Pizza Place   
8              Biggie’s Pizza Riverside           Pizza Place   
9                      Royal Pizza Plus    Italian Restaurant   
10         Joe's & Nona's Pasta & Pizza           Pizza Place   
11                         Tony's Pizza           Pizza Place   
12                            Pizza Hot           Pizza Place   
13                       Pie Five Pizza           Pizza Place   
14                      Doughboys Pizza           Pizza Place   
15                     City of Columbus                  City   
16                        Pizza Rustica           Pizza Place   
17                    Clever Crow Pizza           Pizza Place   
18                          Pizza Shack           Pizza Place   
19                            Pizza Hut           Pizza Place   
20                 Uncle Maddio's Pizza           Pizza Place   
21                     Bisonte Pizza Co           Pizza Place   
22               Brixx Wood Fired Pizza           Pizza Place   
23               Zablong Peculiar Pizza           Pizza Place   
24                      Fuel Pizza Cafe           Pizza Place   
25                Patxi's Chicago Pizza           Pizza Place   
26                 Piccolo Italia Pizza           Pizza Place   
27                 Pizza Zone 'N' Grill           Pizza Place   
28                         Irving Pizza           Pizza Place   
29                         Napoli Pizza           Pizza Place   
...                                 ...                   ...   
6719                       DiFara Pizza           Pizza Place   
6720               Little Caesars Pizza           Pizza Place   
6721                      Marco's Pizza           Pizza Place   
6722               Little Caesars Pizza           Pizza Place   
6723               Little Caesars Pizza           Pizza Place   
6724  Aurelio's Pizza - Las Vegas North           Pizza Place   
6725    John's Incredible Pizza Company           Pizza Place   
6726           Payless Pizza 2 and Ribs           Pizza Place   
6727               Little Caesars Pizza           Pizza Place   
6728                     Domino's Pizza           Pizza Place   
6729                 Peter Pipers Pizza           Pizza Place   
6730              Naked City Pizza Shop           Pizza Place   
6731                Joey's Pizza Nellis           Pizza Place   
6732                    Anthony's Pizza           Pizza Place   
6733               Joe’s New York Pizza           Pizza Place   
6734    Sammy's Woodfired Pizza & Grill           Pizza Place   
6735                          Pizza Hut           Pizza Place   
6736                        Metro Pizza           Pizza Place   
6737                    Slice Bar Pizza           Pizza Place   
6738                      Scili's Pizza           Pizza Place   
6739       Dj Hollywood's Pizza & Gyros  Fast Food Restaurant   
6740               Little Caesars Pizza           Pizza Place   
6741                  Pizza Hut Express           Pizza Place   
6742                    Verrazano Pizza           Pizza Place   
6743                       Jokers Pizza           Pizza Place   
6744               Little Caesars Pizza           Pizza Place   
6745                 Good 2 Go NY Pizza    Italian Restaurant   
6746                     Rosati's Pizza           Pizza Place   
6747               Little Caesars Pizza           Pizza Place   
6748               Little Caesars Pizza           Pizza Place   

    

In [96]:
dfPizza_filtered.groupby('city').count()

name  categories  address   cc  country  crossStreet  \
city                                                                      
AZ                    2           0        2    2        2            2   
Alameda               8           8        8    8        8            2   
Albany                4           4        4    4        4            4   
Albuquerque         103         103       92  103      103           26   
Alcorcón              1           1        1    1        1            0   
Alexandria            7           7        7    7        7            0   
Alhambra              3           3        3    3        3            2   
Allen                 3           3        3    3        3            2   
Altamonte Springs     1           1        1    1        1            0   
Amherst               3           3        3    3        3            1   
Amiens                4           4        4    4        4            0   
Anaheim              56          56       54   56       56           26   
Anchorage            54          54       52   54       54           12   
Antioch               8           8        8    8        8            0   
Apex                  5           5        5    5        5            2   
Apollo Beach          4           4        4    4        4            0   
Apopka                2           2        2    2        2            1   
Aranjuez              2           2        0    2        2            0   
Arcadia               3           3        3    3        3            3   
Archdale              3           3        3    3        3            1   
Argenteuil            1           1        1    1        1            0   
Arlington            51          51       51   51       51           21   
Arroyomolinos         1           1        0    1        1            0   
Arvada                2           2        2    2        2            2   
Athens               19          19       16   19       19            0   
Atlanta              76          76       74   76       76           42   
Aurora                2           2        2    2        2            0   
Austin               99          97       86   99       99           45   
Avon                  2           2        2    2        2            0   
Azle                  2           2        2    2        2            0   
...                 ...         ...      ...  ...      ...          ...   
Washington DC         4           4        4    4        4            2   
Watauga               8           8        4    8        8            4   
Waunakee              1           1        1    1        1            0   
Wauwatosa             2           2        2    2        2            0   
Webster Groves        1           1        1    1        1            0   
West Allis            4           4        2    4        4            2   
West Hollywood        2           2        2    2        2            2   
West Lake Hills       4           4        4    4        4            2   
West Memphis          2           2        2    2        2            0   
West Miami            2           2        2    2        2            0   
West Sacramento       8           8        8    8        8            2   
West Seneca           6           6        6    6        6            2   
Wheaton               2           2        2    2        2            0   
Whitehall             2           2        2    2        2            0   
Whitehouse            2           2        2    2        2            0   
Whittier              4           4        4    4        4            0   
Wichita              95          95       89   95       95           25   
Windsor               6           6        6    6        6            0   
Winter Park           1           1        1    1        1            0   
Wyandotte             2           2        2    2        2            0   
Wyandotte County      2           2        2

In [97]:
dfPizza_Count = dfPizza_filtered.groupby('city').count()
dfPizza_Count

name  categories  address   cc  country  crossStreet  \
city                                                                      
AZ                    2           0        2    2        2            2   
Alameda               8           8        8    8        8            2   
Albany                4           4        4    4        4            4   
Albuquerque         103         103       92  103      103           26   
Alcorcón              1           1        1    1        1            0   
Alexandria            7           7        7    7        7            0   
Alhambra              3           3        3    3        3            2   
Allen                 3           3        3    3        3            2   
Altamonte Springs     1           1        1    1        1            0   
Amherst               3           3        3    3        3            1   
Amiens                4           4        4    4        4            0   
Anaheim              56          56       54   56       56           26   
Anchorage            54          54       52   54       54           12   
Antioch               8           8        8    8        8            0   
Apex                  5           5        5    5        5            2   
Apollo Beach          4           4        4    4        4            0   
Apopka                2           2        2    2        2            1   
Aranjuez              2           2        0    2        2            0   
Arcadia               3           3        3    3        3            3   
Archdale              3           3        3    3        3            1   
Argenteuil            1           1        1    1        1            0   
Arlington            51          51       51   51       51           21   
Arroyomolinos         1           1        0    1        1            0   
Arvada                2           2        2    2        2            2   
Athens               19          19       16   19       19            0   
Atlanta              76          76       74   76       76           42   
Aurora                2           2        2    2        2            0   
Austin               99          97       86   99       99           45   
Avon                  2           2        2    2        2            0   
Azle                  2           2        2    2        2            0   
...                 ...         ...      ...  ...      ...          ...   
Washington DC         4           4        4    4        4            2   
Watauga               8           8        4    8        8            4   
Waunakee              1           1        1    1        1            0   
Wauwatosa             2           2        2    2        2            0   
Webster Groves        1           1        1    1        1            0   
West Allis            4           4        2    4        4            2   
West Hollywood        2           2        2    2        2            2   
West Lake Hills       4           4        4    4        4            2   
West Memphis          2           2        2    2        2            0   
West Miami            2           2        2    2        2            0   
West Sacramento       8           8        8    8        8            2   
West Seneca           6           6        6    6        6            2   
Wheaton               2           2        2    2        2            0   
Whitehall             2           2        2    2        2            0   
Whitehouse            2           2        2    2        2            0   
Whittier              4           4        4    4        4            0   
Wichita              95          95       89   95       95           25   
Windsor               6           6        6    6        6            0   
Winter Park           1           1        1    1        1            0   
Wyandotte             2           2        2    2        2            0   
Wyandotte County      2           2        2

In [98]:
dfPizza_Count.apply(pd.value_counts)
dfPizza_Count

name  categories  address   cc  country  crossStreet  \
city                                                                      
AZ                    2           0        2    2        2            2   
Alameda               8           8        8    8        8            2   
Albany                4           4        4    4        4            4   
Albuquerque         103         103       92  103      103           26   
Alcorcón              1           1        1    1        1            0   
Alexandria            7           7        7    7        7            0   
Alhambra              3           3        3    3        3            2   
Allen                 3           3        3    3        3            2   
Altamonte Springs     1           1        1    1        1            0   
Amherst               3           3        3    3        3            1   
Amiens                4           4        4    4        4            0   
Anaheim              56          56       54   56       56           26   
Anchorage            54          54       52   54       54           12   
Antioch               8           8        8    8        8            0   
Apex                  5           5        5    5        5            2   
Apollo Beach          4           4        4    4        4            0   
Apopka                2           2        2    2        2            1   
Aranjuez              2           2        0    2        2            0   
Arcadia               3           3        3    3        3            3   
Archdale              3           3        3    3        3            1   
Argenteuil            1           1        1    1        1            0   
Arlington            51          51       51   51       51           21   
Arroyomolinos         1           1        0    1        1            0   
Arvada                2           2        2    2        2            2   
Athens               19          19       16   19       19            0   
Atlanta              76          76       74   76       76           42   
Aurora                2           2        2    2        2            0   
Austin               99          97       86   99       99           45   
Avon                  2           2        2    2        2            0   
Azle                  2           2        2    2        2            0   
...                 ...         ...      ...  ...      ...          ...   
Washington DC         4           4        4    4        4            2   
Watauga               8           8        4    8        8            4   
Waunakee              1           1        1    1        1            0   
Wauwatosa             2           2        2    2        2            0   
Webster Groves        1           1        1    1        1            0   
West Allis            4           4        2    4        4            2   
West Hollywood        2           2        2    2        2            2   
West Lake Hills       4           4        4    4        4            2   
West Memphis          2           2        2    2        2            0   
West Miami            2           2        2    2        2            0   
West Sacramento       8           8        8    8        8            2   
West Seneca           6           6        6    6        6            2   
Wheaton               2           2        2    2        2            0   
Whitehall             2           2        2    2        2            0   
Whitehouse            2           2        2    2        2            0   
Whittier              4           4        4    4        4            0   
Wichita              95          95       89   95       95           25   
Windsor               6           6        6    6        6            0   
Winter Park           1           1        1    1        1            0   
Wyandotte             2           2        2    2        2            0   
Wyandotte County      2           2        2

In [99]:
dfPizza_Count =dfPizza_filtered.groupby('city')["id"].count().reset_index(name="count")
dfPizza_Count


city  count
0                   AZ      2
1              Alameda      8
2               Albany      4
3          Albuquerque    103
4             Alcorcón      1
5           Alexandria      7
6             Alhambra      3
7                Allen      3
8    Altamonte Springs      1
9              Amherst      3
10              Amiens      4
11             Anaheim     56
12           Anchorage     54
13             Antioch      8
14                Apex      5
15        Apollo Beach      4
16              Apopka      2
17            Aranjuez      2
18             Arcadia      3
19            Archdale      3
20          Argenteuil      1
21           Arlington     51
22       Arroyomolinos      1
23              Arvada      2
24              Athens     19
25             Atlanta     76
26              Aurora      2
27              Austin     99
28                Avon      2
29                Azle      2
..                 ...    ...
457      Washington DC      4
458            Watauga      8
459           Waunakee      1
460          Wauwatosa      2
461     Webster Groves      1
462         West Allis      4
463     West Hollywood      2
464    West Lake Hills      4
465       West Memphis      2
466         West Miami      2
467    West Sacramento      8
468        West Seneca      6
469            Wheaton      2
470          Whitehall      2
471         Whitehouse      2
472           Whittier      4
473            Wichita     95
474            Windsor      6
475        Winter Park      1
476          Wyandotte      2
477   Wyandotte County      2
478              Wylie      3
479        Yorba Linda      3
480            Yucaipa      2
481              Yukon      2
482           Yvelines      1
483             Éragny      1
484    Санкт-Петербург     47
485         Сестрорецк      1
486              ‘Aiea      2

[487 rows x 2 columns]

In [100]:
dfMediumCityCount = dfMediumCity.join(dfPizza_Count.set_index('city'), on='CityName')
dfMediumCityCount

Number          CityName  Population   Latitude   Longitude  count
10    11
            Austin    978908.0  30.271129  -97.743700   99.0
11    12
      Jacksonville    911507.0  30.332184  -81.655651  105.0
12    13
        Fort Worth    909585.0  32.753177  -97.332746   69.0
13    14
          Columbus    898553.0  39.962260  -83.000707   84.0
14    15
         Charlotte    885708.0  35.227200  -80.843083  103.0
15    16
     San Francisco    881549.0  37.779026 -122.419906  141.0
16    17
      Indianapolis    876384.0  39.768333  -86.158350   97.0
17    18
           Seattle    753675.0  47.603832 -122.330062  103.0
18    19
            Denver    727211.0  39.739236 -104.984862   95.0
19    20
        Washington    705749.0  38.894985  -77.036571   77.0
20    21
            Boston    692600.0  42.360253  -71.058291   91.0
21    22
           El Paso    681728.0  31.775415 -106.464635   79.0
22    23
         Nashville    670820.0  36.162230  -86.774353   71.0
23    24
           Detroit    670031.0  42.331551  -83.046640   67.0
24    25
     Oklahoma City    655057.0  35.472989  -97.517054   89.0
25    26
          Portland    654741.0  45.520247 -122.674195  107.0
26    27
         Las Vegas    651319.0  36.167256 -115.148516  134.0
27    28
           Memphis    651073.0  35.149022  -90.051629   93.0
28    29
        Louisville    617638.0  38.254238  -85.759407   83.0
29    30
         Baltimore    593490.0  39.290882  -76.610759   89.0
30    31
         Milwaukee    590157.0  43.034993  -87.922497   87.0
31    32
       Albuquerque    560513.0  35.084103 -106.650985  103.0
32    33
            Tucson    548073.0  32.222876 -110.974848  105.0
33    34
            Fresno    531576.0  36.729529 -119.708861   77.0
34    35
              Mesa    518012.0  33.415112 -111.831477   64.0
35    36
        Sacramento    513624.0  38.581572 -121.494400   93.0
36    37
           Atlanta    506811.0  33.749099  -84.390185   76.0
37    38
       Kansas City    495327.0  39.100105  -94.578142   61.0
38    39
  Colorado Springs    478221.0  38.833958 -104.825349  101.0
39    40
             Omaha    478192.0  41.258746  -95.938376   83.0
..    ...               ...         ...        ...         ...    ...
59    60
         Lexington    323152.0  38.046407  -84.497039  105.0
60    61
         Henderson    320189.0  32.182598  -95.789318    NaN
61    62
          Stockton    312697.0  37.957702 -121.290780   37.0
62    63
        Saint Paul    308096.0  49.431805    2.010002    2.0
63    64
        Cincinnati    303940.0  39.101454  -84.512460   40.0
64    65
         St. Louis    300576.0  38.626804  -90.199410    NaN
65    66
        Pittsburgh    300286.0  40.441694  -79.990086   54.0
66    67
        Greensboro    296710.0  36.072635  -79.791975   43.0
67    68
           Lincoln    289102.0  47.829907 -118.414727    NaN
68    69
         Anchorage    288000.0  61.216313 -149.894852   54.0
69    70
             Plano    287677.0  33.013676  -96.692510   21.0
70    71
           Orlando    287442.0  28.542111  -81.379030   42.0
71    72
            Irvine    287401.0  33.685697 -117.825982   26.0
72    73
            Newark    282011.0  40.735657  -74.172367   22.0
73    74
            Durham    278993.0  35.996653  -78.901805   32.0
74    75
       Chula Vista    274492.0  32.640054 -117.084196   17.0
75    76
            Toledo    272779.0  39.856068   -4.023957   11.0
76    77
        Fort Wayne    270402.0  41.079990  -85.138601   53.0
77    78
    St. Petersburg    265351.0  59.960674   30.158655    NaN
78    79
            Laredo    262491.0  27.519984  -99.495376   40.0
79    80
       Jersey City    262075.0  40.728158  -74.077642   21.0
80    81
          Chandler    261165.0  33.306222 -111.841332   59.0
81    82
           Madison    259680.0  43.074761  -89.383761   47.0
82    83
           Lubbock    258862.0  33.563521 -101.879336   55.0
83    84
        Scottsdale    258069.0  33.494219 -111.926018   36.0
84    85
    

## Step 4. Calculate the Per Capita Pizza Spots per city

In [101]:
dfMediumCityCount['perCapita'] = dfMediumCityCount['count']/dfMediumCityCount['Population']
dfMediumCityCount

Number          CityName  Population   Latitude   Longitude  count  \
10    11
            Austin    978908.0  30.271129  -97.743700   99.0   
11    12
      Jacksonville    911507.0  30.332184  -81.655651  105.0   
12    13
        Fort Worth    909585.0  32.753177  -97.332746   69.0   
13    14
          Columbus    898553.0  39.962260  -83.000707   84.0   
14    15
         Charlotte    885708.0  35.227200  -80.843083  103.0   
15    16
     San Francisco    881549.0  37.779026 -122.419906  141.0   
16    17
      Indianapolis    876384.0  39.768333  -86.158350   97.0   
17    18
           Seattle    753675.0  47.603832 -122.330062  103.0   
18    19
            Denver    727211.0  39.739236 -104.984862   95.0   
19    20
        Washington    705749.0  38.894985  -77.036571   77.0   
20    21
            Boston    692600.0  42.360253  -71.058291   91.0   
21    22
           El Paso    681728.0  31.775415 -106.464635   79.0   
22    23
         Nashville    670820.0  36.162230  -86.774353   71.0   
23    24
           Detroit    670031.0  42.331551  -83.046640   67.0   
24    25
     Oklahoma City    655057.0  35.472989  -97.517054   89.0   
25    26
          Portland    654741.0  45.520247 -122.674195  107.0   
26    27
         Las Vegas    651319.0  36.167256 -115.148516  134.0   
27    28
           Memphis    651073.0  35.149022  -90.051629   93.0   
28    29
        Louisville    617638.0  38.254238  -85.759407   83.0   
29    30
         Baltimore    593490.0  39.290882  -76.610759   89.0   
30    31
         Milwaukee    590157.0  43.034993  -87.922497   87.0   
31    32
       Albuquerque    560513.0  35.084103 -106.650985  103.0   
32    33
            Tucson    548073.0  32.222876 -110.974848  105.0   
33    34
            Fresno    531576.0  36.729529 -119.708861   77.0   
34    35
              Mesa    518012.0  33.415112 -111.831477   64.0   
35    36
        Sacramento    513624.0  38.581572 -121.494400   93.0   
36    37
           Atlanta    506811.0  33.749099  -84.390185   76.0   
37    38
       Kansas City    495327.0  39.100105  -94.578142   61.0   
38    39
  Colorado Springs    478221.0  38.833958 -104.825349  101.0   
39    40
             Omaha    478192.0  41.258746  -95.938376   83.0   
..    ...               ...         ...        ...         ...    ...   
59    60
         Lexington    323152.0  38.046407  -84.497039  105.0   
60    61
         Henderson    320189.0  32.182598  -95.789318    NaN   
61    62
          Stockton    312697.0  37.957702 -121.290780   37.0   
62    63
        Saint Paul    308096.0  49.431805    2.010002    2.0   
63    64
        Cincinnati    303940.0  39.101454  -84.512460   40.0   
64    65
         St. Louis    300576.0  38.626804  -90.199410    NaN   
65    66
        Pittsburgh    300286.0  40.441694  -79.990086   54.0   
66    67
        Greensboro    296710.0  36.072635  -79.791975   43.0   
67    68
           Lincoln    289102.0  47.829907 -118.414727    NaN   
68    69
         Anchorage    288000.0  61.216313 -149.894852   54.0   
69    70
             Plano    287677.0  33.013676  -96.692510   21.0   
70    71
           Orlando    287442.0  28.542111  -81.379030   42.0   
71    72
            Irvine    287401.0  33.685697 -117.825982   26.0   
72    73
            Newark    282011.0  40.735657  -74.172367   22.0   
73    74
            Durham    278993.0  35.996653  -78.901805   32.0   
74    75
       Chula Vista    274492.0  32.640054 -117.084196   17.0   
75    76
            Toledo    272779.0  39.856068   -4.023957   11.0   
76    77
        Fort Wayne    270402.0  41.079990  -85.138601   53.0   
77    78
    St. Petersburg    265351.0  59.960674   30.158655    NaN   
78    79
            Laredo    262491.0  27.519984  -99.495376   40.0   
79    80
       Jersey City    262075.0  40.728158  -74.077642   21.0   
80    81
          Chandler    261165.0  33.306222 -111.841332   59.0   
81    82
           Madison    259680.0  43.074761  -89.38376

In [102]:
dfMediumCityCount.sort_values(by=['perCapita'],ascending=False)


Number          CityName  Population   Latitude   Longitude  count  \
59    60
         Lexington    323152.0  38.046407  -84.497039  105.0   
58    59
    Corpus Christi    326586.0  27.747725  -97.401413  101.0   
55    56
          Honolulu    345064.0  21.304547 -157.855676   91.0   
51    52
       Bakersfield    384145.0  35.373871 -119.019464  101.0   
49    50
       New Orleans    390144.0  29.949932  -90.070116   97.0   
50    51
           Wichita    389938.0  37.692236  -97.337545   95.0   
46    47
             Tulsa    401190.0  36.155681  -95.992911   97.0   
45    46
       Minneapolis    429606.0  44.977300  -93.265469  101.0   
43    44
    Virginia Beach    449974.0  36.852984  -75.977418  103.0   
80    81
          Chandler    261165.0  33.306222 -111.841332   59.0   
82    83
           Lubbock    258862.0  33.563521 -101.879336   55.0   
38    39
  Colorado Springs    478221.0  38.833958 -104.825349  101.0   
26    27
         Las Vegas    651319.0  36.167256 -115.148516  134.0   
76    77
        Fort Wayne    270402.0  41.079990  -85.138601   53.0   
32    33
            Tucson    548073.0  32.222876 -110.974848  105.0   
84    85
              Reno    255601.0  39.526121 -119.812658   48.0   
68    69
         Anchorage    288000.0  61.216313 -149.894852   54.0   
52    53
         Cleveland    381009.0  41.505161  -81.693445   71.0   
31    32
       Albuquerque    560513.0  35.084103 -106.650985  103.0   
35    36
        Sacramento    513624.0  38.581572 -121.494400   93.0   
81    82
           Madison    259680.0  43.074761  -89.383761   47.0   
65    66
        Pittsburgh    300286.0  40.441694  -79.990086   54.0   
40    41
           Raleigh    474069.0  35.780398  -78.639099   85.0   
39    40
             Omaha    478192.0  41.258746  -95.938376   83.0   
86    87
           Gilbert    254114.0  33.352763 -111.789037   43.0   
25    26
          Portland    654741.0  45.520247 -122.674195  107.0   
41    42
             Miami    467963.0  25.774266  -80.193659   75.0   
15    16
     San Francisco    881549.0  37.779026 -122.419906  141.0   
54    55
           Anaheim    350365.0  33.834752 -117.911732   56.0   
57    58
         Riverside    331360.0  33.953355 -117.396162   51.0   
..    ...               ...         ...        ...         ...    ...   
34    35
              Mesa    518012.0  33.415112 -111.831477   64.0   
37    38
       Kansas City    495327.0  39.100105  -94.578142   61.0   
61    62
          Stockton    312697.0  37.957702 -121.290780   37.0   
14    15
         Charlotte    885708.0  35.227200  -80.843083  103.0   
21    22
           El Paso    681728.0  31.775415 -106.464635   79.0   
11    12
      Jacksonville    911507.0  30.332184  -81.655651  105.0   
73    74
            Durham    278993.0  35.996653  -78.901805   32.0   
16    17
      Indianapolis    876384.0  39.768333  -86.158350   97.0   
19    20
        Washington    705749.0  38.894985  -77.036571   77.0   
22    23
         Nashville    670820.0  36.162230  -86.774353   71.0   
42    43
        Long Beach    462628.0  33.769016 -118.191605   47.0   
10    11
            Austin    978908.0  30.271129  -97.743700   99.0   
23    24
           Detroit    670031.0  42.331551  -83.046640   67.0   
44    45
           Oakland    433031.0  37.804456 -122.271356   41.0   
13    14
          Columbus    898553.0  39.962260  -83.000707   84.0   
71    72
            Irvine    287401.0  33.685697 -117.825982   26.0   
87    88
          Glendale    252381.0  34.146942 -118.247847   22.0   
79    80
       Jersey City    262075.0  40.728158  -74.077642   21.0   
72    73
            Newark    282011.0  40.735657  -74.172367   22.0   
12    13
        Fort Worth    909585.0  32.753177  -97.332746   69.0   
69    70
             Plano    287677.0  33.013676  -96.692510   21.0   
88    89
   North Las Vegas    251974.0  36.200584 -115.121584   17.0   
74    75
       Chula Vista    274492.0  32.640054 -117.08419

## Based on the perCapita Column, I am moving to Lexington!

## Step 5. Cluster the Pizza locations in to 5 groups to find the highest density location

### Now lets see if I can find a neighborhood to move to

In [103]:
dfPizzaLex= dfPizza_filtered[dfPizza_filtered['city'] == 'Lexington']
dfPizzaLex


name   categories                        address  \
242              La Gourmet Pizza  Pizza Place                  325 W Main St   
243     west coast  Gourmet Pizza  Pizza Place             325 W Main St #140   
244                Pie Five Pizza  Pizza Place        867 S Broadway, Ste 130   
245                   Pizza Magia  Pizza Place                319 S Limestone   
246       Puccini's Pizza & Pasta  Pizza Place             833 Chevy Chase Pl   
2793             La Gourmet Pizza  Pizza Place                  325 W Main St   
2794    west coast  Gourmet Pizza  Pizza Place             325 W Main St #140   
2795               Pie Five Pizza  Pizza Place        867 S Broadway, Ste 130   
2796                  Pizza Magia  Pizza Place                319 S Limestone   
2797      Puccini's Pizza & Pasta  Pizza Place             833 Chevy Chase Pl   
2798          Joe Bologna's Pizza  Pizza Place               120 W Maxwell St   
2799           Mad Mushroom Pizza  Pizza Place                341 S Limestone   
2800   Mancino's Pizza & Grinders  Pizza Place               1590 Leestown Rd   
2801                 Pizza Island         Food               308 Newtown Pike   
2802                    Geo Pizza  Pizza Place        150 Avenue of Champions   
2803                        Pizza  Pizza Place            1225 Tates Creek Rd   
2804                    Pizza Hut  Pizza Place             543 S Limestone St   
2805         Little Caesars Pizza  Pizza Place    135 E New Circle Rd Ste 110   
2806                  Naked Pizza  Pizza Place         561 S Broadway Ste 110   
2807                Donatos Pizza  Pizza Place            265 W New Circle Rd   
2808            Papa John's Pizza  Pizza Place       265A Avenue of Champions   
2809                    Sir Pizza  Pizza Place               2604 Richmond Rd   
2810       Happy's Pizza and Subs         Food                600 N Limestone   
2811              Limestone Pizza  Pizza Place                543 S Limestone   
2812               Bearno's Pizza  Pizza Place               396 Woodland Ave   
2813               Hardwood Pizza  Pizza Place              Woodland Art Fair   
2814   Theresa's Pizza & Grinders  Pizza Place                185 Pasadena Dr   
2815               Domino's Pizza  Pizza Place                 801 Euclid Ave   
2816                    Pizza Hut  Pizza Place           393 Waller Ave Ste 4   
2817                  Jet's Pizza  Pizza Place                1030 S Broadway   
...                           ...          ...                            ...   
5315               Hardwood Pizza  Pizza Place              Woodland Art Fair   
5316   Theresa's Pizza & Grinders  Pizza Place                185 Pasadena Dr   
5317               Domino's Pizza  Pizza Place                 801 Euclid Ave   
5318                    Pizza Hut  Pizza Place           393 Waller Ave Ste 4   
5319                  Jet's Pizza  Pizza Place                1030 S Broadway   
5320           Pie Five Pizza Co.  Pizza Place          3401 Nicholasville Rd   
5321                  Pizza Magia  Pizza Place             1300 Winchester Rd   
5322                 Pizza Shoppe  Pizza Place             1300 Winchester Rd   
5323                    Pizza USA  Pizza Place            507 E New Circle Rd   
5324                Donatos Pizza  Pizza Place                  728 E Main St   
5325      Puccini's Pizza & Pasta  Pizza Place                  3801 Dylan Pl   
5326                    Pizza Hut  Pizza Place  1650 Bryan Station Rd Ste 118   
5327                 Apollo Pizza  Pizza Place               1451 Leestown Rd   
5328      Puccini's Pizza & Pasta  Pizza Place              3090 Helmsdale Pl   
5329                    Sir Pizza  Pizza Place               3429 Buckhorn Dr   
5330                   Jets Pizza  Pizza Place                  4222 Saron Dr   
5331                    Sir Pizza  Pizza Place                1076 S Broadway   
5332             LA Gourmet Pizza  Pizza Place              

### I now have my filtered list of Pizza shops in Lexington.  I will now plot them to see them on a map.

In [104]:
address = 'Lexington, KY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Lexington are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Lexington are 38.0464066, -84.4970393.


In [105]:
# create map of New York using latitude and longitude values
map_Lex = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, name, address in zip(dfPizzaLex['lat'], dfPizzaLex['lng'], dfPizzaLex['name'], dfPizzaLex['address']):
    label = '{}, {}'.format(name, address)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Lex)  
    
map_Lex

### Cluster Pizza Locations

In [106]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [107]:
# set number of clusters
kclusters = 5

dfPizzaLex_clustering = dfPizzaLex.drop(['name', 'categories', 'address','cc','city','country','crossStreet','distance','formattedAddress','labeledLatLngs','postalCode','state','id'], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(dfPizzaLex_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [108]:
# add clustering labels
dfPizzaLex.insert(0, 'Cluster Labels', kmeans.labels_)

dfPizzaLex.head() # check the last columns!

Cluster Labels                       name   categories  \
242               1           La Gourmet Pizza  Pizza Place   
243               1  west coast  Gourmet Pizza  Pizza Place   
244               1             Pie Five Pizza  Pizza Place   
245               1                Pizza Magia  Pizza Place   
246               1    Puccini's Pizza & Pasta  Pizza Place   

                     address  cc       city        country crossStreet  \
242            325 W Main St  US  Lexington  United States         NaN   
243       325 W Main St #140  US  Lexington  United States         NaN   
244  867 S Broadway, Ste 130  US  Lexington  United States         NaN   
245          319 S Limestone  US  Lexington  United States         NaN   
246       833 Chevy Chase Pl  US  Lexington  United States         NaN   

     distance                                   formattedAddress  \
242     374.0  [325 W Main St, Lexington, KY 40507, United St...   
243     314.0  [325 W Main St #140, Lexington, KY 40507, Unit...   
244    1727.0  [867 S Broadway, Ste 130, Lexington, KY 40504,...   
245     573.0  [319 S Limestone, Lexington, KY 40508, United ...   
246    2046.0  [833 Chevy Chase Pl, Lexington, KY 40502, Unit...   

                                        labeledLatLngs        lat        lng  \
242  [{'label': 'display', 'lat': 38.04895001941646...  38.048950 -84.499832   
243  [{'label': 'display', 'lat': 38.048645, 'lng':...  38.048645 -84.499229   
244  [{'label': 'display', 'lat': 38.03973, 'lng': ...  38.039730 -84.514829   
245  [{'label': 'entrance', 'lat': 38.043043, 'lng'...  38.043167 -84.502121   
246  [{'label': 'display', 'lat': 38.02868022157393...  38.028680 -84.490857   

    postalCode state                        id  
242      40507    KY  4c02d07c187ec9285ccab47b  
243      40507    KY  51730305498ece19a66bdca0  
244      40504    KY  574341eb498e70237a6738b9  
245      40508    KY  4f32569119836c91c7cde760  
246      40502    KY  4b928c57f964a5202e0334e3

In [109]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [114]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=13)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(dfPizzaLex['lat'], dfPizzaLex['lng'], dfPizzaLex['name'], dfPizzaLex['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [111]:
dfPizzaLex.groupby('Cluster Labels').count()


name  categories  address  cc  city  country  crossStreet  \
Cluster Labels                                                              
0                 20          20       20  20    20       20            0   
1                 59          59       59  59    59       59            8   
2                 10          10       10  10    10       10            4   
3                  8           8        8   8     8        8            2   
4                  8           8        8   8     8        8            0   

                distance  formattedAddress  labeledLatLngs  lat  lng  \
Cluster Labels                                                         
0                     20                20              20   20   20   
1                     59                59              59   59   59   
2                     10                10              10   10   10   
3                      8                 8               8    8    8   
4                      8                 8               8    8    8   

                postalCode  state  id  
Cluster Labels                         
0                       20     20  20  
1                       57     59  59  
2                       10     10  10  
3                        8      8   8  
4                        8      8   8

In [112]:
#Find the Center of the Clusters
k_means_cluster_centers = kmeans.cluster_centers_
k_means_cluster_centers

array([[ 38.05804447, -84.4676803 ],
       [ 38.04341597, -84.50643466],
       [ 37.99456138, -84.5439353 ],
       [ 37.97353283, -84.48902198],
       [ 38.01083955, -84.43551451]])

## The Cluster with the Most Pizza Locations is 1 with 59

## The center of Cluster 1 is 38.04341597, -84.50643466 so I will look for a new home here!!!!